doc/source/train/user_guide.rst

```diff
@@ -532,36 +532,100 @@ workers to be saved on the ``Trainer`` (where your python script is executed).
 The latest saved checkpoint can be accessed through the ``Trainer``'s
 ``latest_checkpoint`` attribute.
 
-As a trivial example, let's create a "model" which is an integer value that is
-equivalent to the sum of the number of epochs.
+Concrete examples are provided to demonstrate how checkpoints are saved
+appropriately in distributed training.
 
-.. code-block:: python
+.. tabs::
 
-    from ray import train
-    from ray.train import Trainer
+  .. group-tab:: PyTorch
 
-    def train_func(config):
-        model = 0 # This should be replaced with a real model.
-        for epoch in range(config["num_epochs"]):
-            model += epoch # This should be replaced with training logic.
-            train.save_checkpoint(epoch=epoch, model=model)
+    .. code-block:: python
 
-    trainer = Trainer(backend="torch", num_workers=2)
-    trainer.start()
+        import ray.train.torch
+        from ray import train
+        from ray.train import Trainer
+
+        import torch
+        import torch.nn as nn
+        from torch.nn.modules.utils import consume_prefix_in_state_dict_if_present
+        from torch.optim import Adam
+        import numpy as np
+
+
+        def train_func(config):
+            n = 100
+            # create a toy dataset
+            # data   : X - dim = (n, 4)
+            # target : Y - dim = (n, 1)
+            Y = torch.Tensor(np.random.uniform(0, 1, size=(n, 1)))
+            X = torch.Tensor(np.random.normal(0, 1, size=(n, 4)))
+            # toy neural network : 1-layer
+            # wrap the model in DDP
+            model = ray.train.torch.prepare_model(nn.Linear(4, 1))
+            mse = nn.MSELoss()
+
+            optimizer = Adam(model.parameters(), lr=3e-4)
+            for epoch in range(config["num_epochs"]):
+                y = model.forward(X)
+                # compute loss
+                loss = mse(y, Y)
+                # back-propagate loss
+                optimizer.zero_grad()
+                loss.backward()
+                optimizer.step()
+                # To fetch non-DDP state_dict
+                # w/o DDP: model.state_dict()
+                # w/  DDP: model.module.state_dict()
+                # See: https://github.com/ray-project/ray/issues/20915
+                ddp_state_dict = model.state_dict()
```

Comment:
nit: Rename this to `state_dict` since this should also work if the model is not wrapped in DDP (when `num_workers=1`).